keps/sig-node/4569-cgroup-v1-deprecation/README.md

```diff
@@ -0,0 +1,677 @@
+
+#### Introduction of DeprecatedCgroupV1 Feature Gate
+
+With the introduction of the `DeprecatedCgroupV1` feature gate, Kubernetes aims to guide users through the transition from cgroup v1 to cgroup v2 by providing clear and timely warnings. When this feature gate is enabled, kubelet will log a warning message in scenarios where it detects the host system is utilizing cgroup v1. This approach ensures users are informed of the deprecation and encourages planning for migration to cgroup v2.
+
+
+#### Code Modifications for Warning Log Implementation
+
+To incorporate these warning logs, the kubelet's code will be adjusted in locations where the IsCgroup2UnifiedMode() function is invoked. e.g.
+
+Original Code Snippet:
+```golang
+if m.memorySwapBehavior == kubelettypes.LimitedSwap {
+		if !isCgroup2UnifiedMode() {
+			swapConfigurationHelper.ConfigureNoSwap(lcr)
+			return
+		}
+	}
+```
+Revised Code Snippet:
+```golang
+if m.memorySwapBehavior == kubelettypes.LimitedSwap {
+    if !isCgroup2UnifiedMode() && utilfeature.DefaultFeatureGate.Enabled(features.DeprecatedCgroupV1) {
+      klog.Warning("cgroup v1 support has been deprecated, please plan for the migration towards cgroup v2")
+      swapConfigurationHelper.ConfigureNoSwap(lcr)
+      return
+    }
+}
+```
+
+Furthermore, code segments that default to cgroup v1 logic will be inverted to assume cgroup v2 as the default. This shift underscores the transition to cgroup v2 and ensures warning logs are consistently generated for cgroup v1 usage. e.g.
+
+Original Code Snippet:
+```golang
+	memLimitFile := "memory.limit_in_bytes"
+	if libcontainercgroups.IsCgroup2UnifiedMode() {
+		memLimitFile = "memory.max"
+	}
+```
+Revised Code Snippet:
+```golang
+	memLimitFile = "memory.max"
+	if libcontainercgroups.!IsCgroup2UnifiedMode() && utilfeature.DefaultFeatureGate.Enabled(features.DeprecatedCgroupV1) {
+      klog.Warning("cgroup v1 support has been deprecated, please plan for the migration towards cgroup v2")
+      memLimitFile := "memory.limit_in_bytes"
+	}
+```
+
+### Feature Flag and Kubelet Option
+
+During the `alpha` stage of this feature's lifecycle, it's anticipated that users will require time to plan and execute their migrations to cgroup v2. Factors influencing the migration timeline include, but are not limited to, selecting compatible kernels and operating systems, planning for potential downtimes, and coordinating changes to minimize disruption during peak operational periods.
+
+To accommodate this, we propose introducing a kubelet flag, `--suppress-cgroupv1-deprecation-warning`, which is not set by default. This flag provides users the flexibility to suppress repetitive deprecation warning logs related to cgroup v1 while they are in the process of planning and executing their migration to cgroup v2. The introduction of this flag is intended to reduce log spam and operational noise, acknowledging the non-trivial nature of the migration for many users.
```

Comment:
I think we should bring this up if we actually want this flag. Is the goal of this flag for just alpha?

once we hit beta/ga of this feature, will we drop this flag?

Reply:
I see the section below! disregard.